### PR TITLE
Add setElementCollidableWithType

### PR DESCRIPTION
**Pull request by emre1702**
With this function you can disable the collision of a element with all the elements of a type and also new created ones.
I had struggles to create this kind of script in my newbie-times and also experienced many other new scripters having problems. So why not create a simple function which does, what you want?

Syntax:
bool setElementCollidableWithType ( element theElement, string withType, bool enabled[, bool onlyWithCreated = false ])

Use "onlyWithCreated" if you don't want to consider new created elements (like it is right now in setElementCollidableWith).

Oh and the first commits could be strange.
Wanted to overload the setElementCollidableWith function, then decided to create a new function.

### CODE REVIEW

File: Client/mods/deathmatch/logic/CClientPlayer.cpp [hunk 1]
```diff
@@ -88,6 +92,16 @@ CClientPlayer::CClientPlayer ( CClientManager* pManager, ElementID ID, bool bIsL
     // Add us to the player list
     m_pManager->GetPlayerManager ()->AddToList ( this );
 
+    // Check DisableCollisions //
+    if ( !m_DisabledCollisions.empty () )
+    {
+        std::list < CClientEntity * > ::iterator iter = m_DisabledCollisions.begin ();
```
Comment (Necktrox) [inline Client/mods/deathmatch/logic/CClientPlayer.cpp]: You added `using std::list;` on top of this translation unit, but kept the **std::** namespace here

File: Client/mods/deathmatch/logic/CClientPlayer.cpp [hunk 2]
```diff
@@ -88,6 +92,16 @@ CClientPlayer::CClientPlayer ( CClientManager* pManager, ElementID ID, bool bIsL
...
+        for ( ; iter != m_DisabledCollisions.end (); iter++ )
```
Comment (Necktrox) [inline Client/mods/deathmatch/logic/CClientPlayer.cpp]: You can use a for-range loop here

File: Client/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp [hunk 3]
```diff
@@ -2180,6 +2182,33 @@ int CLuaElementDefs::SetElementCollidableWith ( lua_State* luaVM )
...
+    SString strType;
+    bool bCanCollide = true;
+    bool bOnlyWithCreated = false;
+    CScriptArgReader argStream ( luaVM );
+    argStream.ReadUserData ( pEntity );
+    argStream.ReadString ( strType );
+    argStream.ReadBool ( bCanCollide );
+    argStream.ReadBool ( bOnlyWithCreated, false );
+
+    // Verify the arguments
```
Comment (Necktrox) [inline Client/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp]: You might have copy-pasted this, but the comment doesn't serve any important information, which wasn't clear beforehand.

File: Client/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp [hunk 4]
```diff
@@ -2180,6 +2182,33 @@ int CLuaElementDefs::SetElementCollidableWith ( lua_State* luaVM )
 }
 
 
+int CLuaElementDefs::SetElementCollidableWithType ( lua_State* luaVM ) {
+    // bool SetElementCollidableWithType ( element theElement, string withType, bool enabled, bool onlyWithCreated = false )
```
Comment (Necktrox) [inline Client/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp]: It should begin with a lower-case **s** here and the optional argument should be surrounded with [ ] brackets:
`bool setElementCollidableWithType ( element theElement, string withType, bool enabled [, bool onlyWithCreated = false  ] )`

File: Client/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp [hunk 5]
```diff
@@ -2180,6 +2182,33 @@ int CLuaElementDefs::SetElementCollidableWith ( lua_State* luaVM )
 }
 
 
+int CLuaElementDefs::SetElementCollidableWithType ( lua_State* luaVM ) {
+    // bool SetElementCollidableWithType ( element theElement, string withType, bool enabled, bool onlyWithCreated = false )  
+    CClientEntity* pEntity = NULL;
```
Comment (Necktrox) [inline Client/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp]: Use **nullptr**

File: Client/mods/deathmatch/logic/CClientPed.cpp [hunk 6]
```diff
@@ -242,6 +243,16 @@ void CClientPed::Init ( CClientManager* pManager, unsigned long ulModelID, bool
 
         SetArmor ( 0.0f );
     }
+
+    // Check DisableCollisions //
+    if ( !m_DisabledCollisions.empty () )
+    {
+        list < CClientEntity * > ::iterator iter = m_DisabledCollisions.begin ();
+        for ( ; iter != m_DisabledCollisions.end (); iter++ )
```
Comment (Necktrox) [inline Client/mods/deathmatch/logic/CClientPed.cpp]: Same applies here, you can use a for-range loop

File: Client/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp [hunk 7]
```diff
@@ -3603,6 +3603,77 @@ bool CStaticFunctionDefinitions::SetElementCollidableWith ( CClientEntity & Enti
...
+                case CCLIENTPLAYER:
+                case CCLIENTPED:
+                case CCLIENTOBJECT:
+                case CCLIENTVEHICLE:
+                {
+                    // Change CollidableWith for all elements of this type on the server
+                    CFastList < CClientEntity* > entities = CClientEntity::GetEntitiesFromRoot ( HashString ( szTypeName ), false );
+
+                    CChildListType::const_iterator iter = entities.begin ();
+                    for ( ; iter != entities.end (); iter++ )
```
Comment (Necktrox) [inline Client/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp]: I am not certain if you can use a for-range loop here, but **.begin()** and **.end()** seem promising

**Comment by emre1702** [comment]
Changed all, thanks for the infos (C++ newbie here ^^)

File: Client/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp (hunk 7 above)
Comment (emre1702) [reply Client/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp]: Should I change GetEntitiesFromRoot return to list instead of CFastList?

File: Client/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp [hunk 8]
```diff
@@ -3603,6 +3603,77 @@ bool CStaticFunctionDefinitions::SetElementCollidableWith ( CClientEntity & Enti
 }
 
 
+bool CStaticFunctionDefinitions::SetElementCollidableWithType ( CClientEntity & Entity, const char* szTypeName, bool bCanCollide, bool bOnlyWithCreated )
+{
+    switch ( Entity.GetType () )
```
Comment (Necktrox) [inline Client/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp:3608]: You have a few nested code blocks here, but you could do **return false;** in the **default** case and do **break** for the supported entity types and continue after the switch.

File: Client/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp (hunk 7 above)
Comment (Necktrox) [reply Client/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp]: You should rather wait for somebody else to comment or benchmark both variants and keep the better one, because I can't answer this question.

File: Client/mods/deathmatch/logic/CClientPed.cpp [hunk 9]
```diff
@@ -16,10 +16,10 @@
 
 #include "StdInc.h"
 
-using std::list;
```
Comment (Necktrox) [inline Client/mods/deathmatch/logic/CClientPed.cpp:19]: Undo this change to fix the compilation errors.

File: Client/mods/deathmatch/logic/CClientVehicle.cpp [hunk 10]
```diff
@@ -21,10 +21,9 @@
 
 #include "StdInc.h"
 
-using std::list;
```
Comment (Necktrox) [inline Client/mods/deathmatch/logic/CClientVehicle.cpp]: Undo this change to fix the compilation errors.

File: Client/mods/deathmatch/logic/CClientEntity.cpp [hunk 11]
```diff
@@ -175,19 +175,10 @@ CClientEntity::~CClientEntity ( void )
     g_pCore->UpdateDummyProgress();
 
     // Remove from m_DisabledCollisions
-    switch ( this->GetType () )
```
Comment (emre1702) [inline Client/mods/deathmatch/logic/CClientEntity.cpp]: Does someone understand why this line (without "this->" too) gives a "LNK2019" error?
Maybe because it's CClientVehicle is getting destructed before, so GetType is not implemented for the element anymore?

Comment (sbx320) [reply Client/mods/deathmatch/logic/CClientEntity.cpp]: Calling any virtual member function in a destructor is a very bad idea.

Comment (Necktrox) [reply Client/mods/deathmatch/logic/CClientEntity.cpp]: GetType is virtual

Comment (emre1702) [reply Client/mods/deathmatch/logic/CClientEntity.cpp]: So either not check the type and do it for all element-types or get the type in another way.

**Comment by Einheit-101** [comment]
Is it possible to enhance this in a way so that its possible to make an element non-collidable with WORLD elements? This would allow vehicles to pass through certain default world objects.

**Comment by qaisjp** [comment]
A version of this PR with master merged in can be found here: https://github.com/qaisjp/mtasa-blue/tree/setElementCollidableWithImprovement

If someone wants to take it forward, please do so. I can't push to this PR  because @emre1702 deleted their fork.

**Comment by Woovie** [comment]
Replaced with #1719
